Selenium web driver

EugenMakarevich opened the pull request:
Task 40

EugenMakarevich committed:
Selenium WebDriver. Work in progress

EugenMakarevich committed:
Selenium WebDriver. Work in progress

EugenMakarevich committed:
Selenium WebDriver. Work in progress

EugenMakarevich committed:
Selenium WebDriver. Work in progress

EugenMakarevich committed:
Selenium WebDriver. Work in progress

EugenMakarevich committed:
Selenium WebDriver. Final small changes

shukal94 commented on src/test/java/selenium/AlertsTest.java:
```diff
@@ -0,0 +1,77 @@
+        String alertText = alert.getText();
+
+        log.info("Verify the alert text is '{}'", ALERT_TEXT);
+        Assert.assertEquals(alertText, ALERT_TEXT);
```
always use an assertion message for each assertions as an extra parameter, in this case it will be like "Alert text mismatch!"

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/AlertsTest.java:
```diff
@@ -0,0 +1,77 @@
+        WebElement resultElement = driver().findElement(CONFIRM_BOX_RESULT);
+        String resultText = resultElement.getText();
+
+        Assert.assertEquals(resultText, OK_BUTTON_RESULT_TEXT);
```
see above

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/AlertsTest.java:
```diff
@@ -0,0 +1,77 @@
+        WebElement resultElement = driver().findElement(CONFIRM_BOX_RESULT);
+        String resultText = resultElement.getText();
+
+        Assert.assertEquals(resultText, CANCEL_BUTTON_RESULT_TEXT);
```
see above

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/LoadingDataDynamicallyTest.java:
```diff
@@ -0,0 +1,36 @@
+public class LoadingDataDynamicallyTest extends TestBase {
+
+    @Test
+    public void TestLoadingDataDynamically() {
```
follow java coding style conventions

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/LoadingDataDynamicallyTest.java:
```diff
@@ -0,0 +1,36 @@
+public class LoadingDataDynamicallyTest extends TestBase {
+
+    @Test
+    public void TestLoadingDataDynamically() {
```
this test does not contain any validation, we should expect at least one condition which indicates that the test passed

EugenMakarevich replied:
Added 3 assertions. Done!

shukal94 commented on src/test/java/selenium/MultiselectTest.java:
```diff
@@ -0,0 +1,48 @@
+public class MultiselectTest extends TestBase {
+
+    @Test
+    public void testMultiselect() {
```
test multiselect what? we need to state the goal of testing by meaningful name (like testcase name)

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/ProgressBarTest.java:
```diff
@@ -0,0 +1,38 @@
+            }
+        }
+
+        log.info("Refresh the page");
+        driver().navigate().refresh();
```
what do you test here? for future - use only buttons that the application provides, native browser buttons during testing could lead to flaky tests

EugenMakarevich replied:
Added assertion. Done!

shukal94 commented on src/test/java/selenium/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,170 @@
+        }
+        Assert.assertEquals(optionToBeSelected, showEntriesDropdown.getFirstSelectedOption().getText());
+
+        int age = 40;
+        double salary = 150000.0;
```
move to class-level constant

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    public class Employee {
```
1. let's use lombok here
2. move this class to src/main/java/model package (if you don't have such - create this one)
3. getEmployeeDataFromPage, getEmployeeDataFromAllPages, sortEmployees inside of this class violate single-responsibility principle - move those methods to a separate class like Employees

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,170 @@
+
+        log.info("Print employees who have age > {} and salary <= {}", age, salary);
+        for (Employee employee : sortedList) {
+            System.out.println(employee.toString());
```
use log instead

EugenMakarevich replied:
Done!

shukal94 commented on src/test/java/selenium/YandexMailLoginTest.java:
```diff
@@ -1,52 +1,51 @@
-    void setUp() {
-        driver = WebDriverUtils.setWebDriver();
-        wait = new WebDriverWait(driver, SHORT_TIMEOUT);
+    public void setUp1() {
```
setUp1?

EugenMakarevich replied:
Done!

shukal94 reviewed (changes requested):
let's follow the standard maven archetype
src/test/java/com/coherentsolutions/aqa/java/web/makarevich packages hierarchy for tests only
the rest of the stuff (constants, pages, models, services, etc.) should be located after src/main/java/com/coherentsolutions/aqa/java/web/makarevich

andrei-unuk commented on src/test/java/parameters/TestData.java:
```diff
@@ -0,0 +1,16 @@
+public class TestData {
+
+    @DataProvider
+    public static Object[][] yandexMailCredentials() {
```
Usualy there is no need to keep datatprovider in a separate class - you can keep it in the same class with test that uses it.

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/selenium/YandexMailLoginTest.java:
```diff
@@ -1,52 +1,51 @@
-        driver.findElement(LOGIN_BUTTON_AUTHORIZATION_PAGE).click();
+        driver().findElement(EMAIL_FIELD).sendKeys(username);
+        driver().findElement(LOGIN_BUTTON_AUTHORIZATION_PAGE).click();
+        Thread.sleep(3000); //3.Add Thread.sleep for login test. What type of waiter is it - Implicit waiter?
```
Are u sure that it is implicit?

EugenMakarevich replied:
It is not implicit nor explicit, since Thread.sleep() does not wait for any specific condition to be met, nor does it interact with the DOM or elements in any way.

andrei-unuk commented on src/test/java/selenium/YandexMailLoginTest.java:
```diff
@@ -1,52 +1,51 @@
-    void tearDown() {
-        if (driver != null) driver.quit();
+        new WebDriverWait(driver(),SHORT_TIMEOUT, Duration.ofMillis(300)).until(ExpectedConditions.visibilityOfElementLocated(USER_ACCOUNT_NAME)); //4. Add explicit waiter for login test, which will wait until name appears (after login). Add polling frequency, which is differ from default value
+        Assert.assertEquals(username, driver().findElement(USER_ACCOUNT_NAME).getText());
```
1. I suggest to use multy assertions in case when you have more than 1 assert in your test
2. I also suggest to create a variables for things like 'driver().findElement(USER_ACCOUNT_NAME).getText()' and 'driver().getTitle().contains(YANDEX_MAIL_INBOX_TITLE)'. It is not a strict ule but it will make your code more reaeble and will make it easier for debugging.

EugenMakarevich replied:
1. Done for all tests
2. Done for all tests

andrei-unuk commented on src/test/java/constants/TestConstants.java:
```diff
@@ -5,9 +5,20 @@
+    public static final String YANDEX_MAIL_PASSWORD_2 = ConfigUtils.getProperty("yandexmail.password_2");
     public static final String CONFIG_PROPERTIES = "src/test/resources/config.properties";
     public static final String YANDEX_MAIL_INBOX_TITLE = "Inbox — Yandex Mail";
+    public static final String MULTISELECT_TEST_URL = "https://demo.seleniumeasy.com/basic-select-dropdown-demo.html";
```
Keeping URLs in property files instead of doing it in classes - is a common practice

EugenMakarevich replied:
Done

andrei-unuk commented on src/test/java/selenium/ProgressBarTest.java:
```diff
@@ -0,0 +1,38 @@
+public class ProgressBarTest extends TestBase {
+
+    @Test
+    public void TestProgressBar() {
```
https://github.com/EugenMakarevich/SeleniumWebDriver/pull/4/files#r1328437655

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/selenium/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,170 @@
+public class TableSortAndSearchTest extends TestBase {
+
+    @Test
+    public void TestTableSortAndSearch() {
```
https://github.com/EugenMakarevich/SeleniumWebDriver/pull/4/files#r1328437655

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/selenium/YandexMailLoginTest.java:
```diff
@@ -1,52 +1,51 @@
-    @Test
-    void TestYandexMailLogin() {
+    @Test(dataProvider = "yandexMailCredentials", dataProviderClass = TestData.class)
+    public void TestYandexMailLogin(String username, String password) throws InterruptedException {
```
https://github.com/EugenMakarevich/SeleniumWebDriver/pull/4/files#r1328437655

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/selenium/ProgressBarTest.java:
```diff
@@ -0,0 +1,38 @@
+public class ProgressBarTest extends TestBase {
+
+    @Test
+    public void TestProgressBar() {
```
https://github.com/EugenMakarevich/SeleniumWebDriver/pull/4/files#r1328438684

EugenMakarevich replied:
Done!

EugenMakarevichCoherent committed:
Selenium WebDriver. Working on the code review comments. In Progress.

EugenMakarevichCoherent committed:
Selenium WebDriver. Working on the code review comments. In Progress.

EugenMakarevichCoherent committed:
Selenium WebDriver. Working on the code review comments. In Progress.

EugenMakarevichCoherent committed:
Selenium WebDriver:
1. Worked on the code review comments. All comments are addressed.
2. Restructure the whole project to follow the standard maven archetype
3. Deleted the obsolete files from previous tasks: the files of application, unit tests and smoke and regression suites.

andrei-unuk commented on src/main/java/com/coherentsolutions/aqa/web/makarevich/model/Employee.java:
```diff
@@ -0,0 +1,27 @@
+    private String position;
+    private String office;
+    private int age;
+    private String start_date;
```
startDate

EugenMakarevich replied:
Done!

andrei-unuk commented on src/main/java/com/coherentsolutions/aqa/web/makarevich/constants/TestConstants.java:
```diff
@@ -0,0 +1,19 @@
+    public static final String LAST_NAME_TEXT = "Last Name";
+    public static final String DEFAULT_PROGRESS_BAR_VALUE = "0%";
+    public static final int DOWNLOAD_PERCENT = 50;
+    public static final double TAX = 0.2;
```
This variable is not used. Please remove it.

EugenMakarevich replied:
Done!

andrei-unuk commented on src/main/java/com/coherentsolutions/aqa/web/makarevich/testbase/TestBase.java:
Please move this class in same package as your test classes are in.

EugenMakarevich replied:
Done

andrei-unuk commented on src/test/java/com/coherentsolutions/aqa/java/web/makarevich/LoadingDataDynamicallyTest.java:
```diff
@@ -0,0 +1,53 @@
+        softAssert.assertTrue(userInfoText.contains(LAST_NAME_TEXT), String.format("%s text is not present in the user information block", LAST_NAME_TEXT));
+        softAssert.assertNotNull(imgTag, "The img tag is not present in the DOM within the user information block");
+        softAssert.assertAll();
+        log.info("User Info: {}", userInfoText);
```
Is there any reason in logging AFTER assertions? If no - please set it before assertions (becose there will not display in case of assert fail)

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/com/coherentsolutions/aqa/java/web/makarevich/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,46 @@
+@Slf4j
+public class TableSortAndSearchTest extends TestBase {
+
+    public static final String optionToBeSelected = "10";
```
OPTIONS_TO_BE_SELECTED

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/com/coherentsolutions/aqa/java/web/makarevich/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,46 @@
+public class TableSortAndSearchTest extends TestBase {
+
+    public static final String optionToBeSelected = "10";
+    public static final int age = 40;
```
AGE

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/com/coherentsolutions/aqa/java/web/makarevich/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    public static final String optionToBeSelected = "10";
+    public static final int age = 40;
+    public static final double salary = 150000.0;
```
SALARY

EugenMakarevich replied:
Done!

andrei-unuk commented on src/test/java/com/coherentsolutions/aqa/java/web/makarevich/TableSortAndSearchTest.java:
```diff
@@ -0,0 +1,46 @@
+        List<Employee> sortedEmployeeList = new Employees(driver()).sortEmployees(age, salary);
+        Assert.assertFalse(sortedEmployeeList.isEmpty(), "Sorted employee list is empty!");
+
+        log.info("Print sorted employees");
```
https://github.com/EugenMakarevich/SeleniumWebDriver/pull/4/files#r1335475676

EugenMakarevich replied:
Deleted that part since it's not obligatory per test case

EugenMakarevichCoherent committed:
Worked on code review comments.
